### PR TITLE
Keyboard

### PR DESCRIPTION
**Pull request by ddalp**
#2852 

- Add support to listen and fire "keyboardDidShow" and "keyboardDidHide" events when SIP is shown or hidden.  Basically hooking up CoreInputView's OcclusionsChanged event.
- Manually tested these two RN events firing and verified with keyboardAvoidingView RN object working fine.

###### Microsoft Reviewers: [Open in CodeFlow](http://wpcp.azurewebsites.net/CodeFlowProtocolProxy2.php?pullrequest=https://github.com/microsoft/react-native-windows/pull/3541)

### CODE REVIEW

File: vnext/ReactUWP/Views/SIPEventHandler.cpp [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+#include <Utils/RectUtils.h>
+
+#include <winrt/Windows.ApplicationModel.Core.h>
+#include <winrt/Windows.Devices.Input.h>
+#include <winrt/Windows.Foundation.h>
+#include <winrt/Windows.UI.Core.h>
+#include <winrt/Windows.UI.Input.h>
+#include <winrt/Windows.UI.Xaml.Controls.h>
+#include <winrt/Windows.UI.Xaml.Input.h>
+#include <winrt/Windows.UI.Xaml.Media.h>
```
Comment (ddalp) [inline vnext/ReactUWP/Views/SIPEventHandler.cpp]: Remove.

File: vnext/include/ReactUWP/Utils/RectUtils.h [hunk 2]
```diff
@@ -0,0 +1,18 @@
...
+#include <ReactWindowsCore/ReactWindowsAPI.h>
+#include <winrt/Windows.Foundation.h>
+
+namespace react {
+namespace uwp {
+
+REACTWINDOWS_API_(bool) IsEmptyRect(winrt::Windows::Foundation::Rect rect);
+REACTWINDOWS_API_(winrt::Windows::Foundation::Rect) CreateEmptyRect();
+REACTWINDOWS_API_(winrt::Windows::Foundation::Rect)
+UnionRect(winrt::Windows::Foundation::Rect rect1, _In_ winrt::Windows::Foundation::Rect rect2);
```
Comment (asklar) [inline vnext/include/ReactUWP/Utils/RectUtils.h]: when passing Rect pass instead const Rect&

File: vnext/ReactUWP/Utils/RectUtils.cpp [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+float GetRight(_In_ winrt::Rect rect) {
+  return (IsEmptyRect(rect) ? NegativeInfinity : rect.X + rect.Width);
+}
+
+// This is a read-only alias for Y + Height.  If this is the empty
+// rectangle, the value will be negative infinity.
+float GetBottom(_In_ winrt::Rect rect) {
+  return (IsEmptyRect(rect) ? NegativeInfinity : rect.Y + rect.Height);
+}
+
```
Comment (asklar) [inline vnext/ReactUWP/Utils/RectUtils.cpp]: not used, remove?

File: vnext/ReactUWP/Utils/RectUtils.cpp [hunk 4]
```diff
@@ -0,0 +1,117 @@
...
+REACTWINDOWS_API_(bool) IsEmptyRect(winrt::Rect rect) {
+  return rect.Width < 0;
```
Comment (asklar) [inline vnext/ReactUWP/Utils/RectUtils.cpp]: >< [](start = 20, length = 1)

<= ?

File: vnext/ReactUWP/Utils/RectUtils.cpp [hunk 5]
```diff
@@ -0,0 +1,117 @@
...
+}
+
+REACTWINDOWS_API_(winrt::Rect) CreateEmptyRect() {
+  winrt::Rect result;
+  result.X = std::numeric_limits<float>::infinity();
+  result.Y = std::numeric_limits<float>::infinity();
+  result.Width = -std::numeric_limits<float>::infinity();
+  result.Height = -std::numeric_limits<float>::infinity();
```
Comment (asklar) [inline vnext/ReactUWP/Utils/RectUtils.cpp]: should set these to 0?

File: vnext/ReactUWP/Utils/RectUtils.cpp [hunk 6]
```diff
@@ -0,0 +1,117 @@
...
+  return result;
+}
+
+// Update this rectangle to be the union of this and rect.
+REACTWINDOWS_API_(winrt::Rect) UnionRect(winrt::Rect rect1, _In_ winrt::Rect rect2) {
+  winrt::Rect result = rect1;
+  if (IsEmptyRect(rect1)) {
```
Comment (asklar) [inline vnext/ReactUWP/Utils/RectUtils.cpp]: can you just call the user32 UnionRect?
UnionRect(ToRECT(rect1), ToRECT(rect2))

File: vnext/ReactUWP/Utils/RectUtils.cpp (hunk 4 above)
Comment (asklar) [reply vnext/ReactUWP/Utils/RectUtils.cpp]: In fact consider reusing user32's IsRectEmpty instead

---
In reply to: [339777475](https://github.com/microsoft/react-native-windows/pull/3541#discussion_r339777475) [](ancestors = 339777475)

File: vnext/ReactUWP/Views/SIPEventHandler.h [hunk 7]
```diff
@@ -0,0 +1,31 @@
...
+
+class SIPEventHandler {
+ public:
+  SIPEventHandler(const std::weak_ptr<IReactInstance> &reactInstance);
+  virtual ~SIPEventHandler();
+
+ private:
+  void SendEvent(std::string &&eventName, folly::dynamic &&parameters);
+  std::weak_ptr<IReactInstance> m_wkReactInstance;
+  winrt::CoreInputView::OcclusionsChanged_revoker m_occlusionsChnaged_revoker;
```
Comment (asklar) [inline vnext/ReactUWP/Views/SIPEventHandler.h]: >m_occlusionsChnaged_revoker [](start = 50, length = 27)

fix typo: Ch **n** aged

File: vnext/include/ReactUWP/Utils/RectUtils.h (hunk 2 above)
Comment (ddalp) [inline vnext/include/ReactUWP/Utils/RectUtils.h]: There is already RectHelper class I can use, will remove .

**Pull request merged**